backend/cellguide/pipeline/ontology_tree/tree_builder.py

```diff
@@ -205,3 +213,332 @@ def _traverse_ontology_with_counting(
+        """
+        This function truncates the ontology graph in-place by removing nodes that are not in the valid nodes list.
+        It also adds a dummy node to the graph if the graph has children that are not in the valid nodes list.
+        The dummy node contains the ids of the invalid children.
+
+        Arguments
+        ---------
+        graph: dict
+            A node of the ontology graph. At the top-level, this should be the root node of the ontology graph.
+        valid_nodes: list
+            A list of valid cell type ontology term ids.
+        """
+        if graph["id"] not in valid_nodes:
+            return False
+
+        children = graph.get("children", [])
+        valid_children = []
+        append_dummy = False
+
+        invalid_children_ids = []
+        for child in children:
+            is_valid = self._truncate_graph_first_pass(child, valid_nodes)
+            if is_valid:
+                valid_children.append(child)
+            elif child["id"] != "":
+                invalid_children_ids.append(child["id"])
+                append_dummy = True
+
+        if append_dummy and len(valid_children) > 0:
+            valid_children.append(
+                {
+                    "id": "",
+                    "name": "",
+                    "n_cells_rollup": 0,
+                    "n_cells": 0,
+                    "invalid_children_ids": invalid_children_ids,
+                    "parent": graph["id"],
+                }
+            )
+        if len(valid_children) > 0:
+            graph["children"] = valid_children
+        else:
+            if "children" in graph:
+                del graph["children"]
+
+        return True
+
+    def _truncate_graph_second_pass(self, graph, visited_nodes_in_paths, nodesWithChildrenFound=None):
+        """
+        We do not want to show cell types multiple times in the ontology tree unless they are nodes that are in one of the
+        paths to the target cell type. In this case, show the node in the path and collapse all its siblings into the
+        dummy node.
+
+        Arguments
+        ---------
+        graph: dict
+            A node of the ontology graph. At the top-level, this should be the root node of the ontology graph.
+        visited_nodes_in_paths: list
+            A list of cell type ontology term ids that are in one of the paths to the target cell type.
+            The target cell type may have multiple instances in the ontology graph (each instance with a unique suffix).
+        nodesWithChildrenFound: set, optional, default=None
+            A set of cell type ontology term ids that have children. This is used to remove children of nodes that
+            have already been encountered during the traversal.
+        """
+        if nodesWithChildrenFound is None:
+            nodesWithChildrenFound = set()
+        if graph["id"].split("__")[0] in nodesWithChildrenFound:
+            if "children" in graph:
```

Comment:
i would add a code comment explaining what we're doing here, which (if im understanding correctly) is collapsing the siblings into a dummy node